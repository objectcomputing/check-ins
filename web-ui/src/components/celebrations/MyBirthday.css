@import url('https://fonts.googleapis.com/css?family=Wendy+One');

body {
    font-family: "Wendy One", sans-serif;
  }

  p {
    text-transform: uppercase;
  }
.my-birthday-container {
    position: relative;
}
.my-birthday h1 {
    color: #2559a7;
    font-weight: bolder;
}
.my-balloons {
    margin: 0 auto;
    position: absolute;
    top: -2rem;
  }
  .my-balloons > div {
    width: 104px;
    height: 140px;
    background: rgba(182, 15, 97, 0.9);
    border-radius: 0;
    border-radius: 80% 80% 80% 80%;
    margin: 0 auto;
    position: absolute;
    padding: 10px;
    box-shadow: inset 17px 7px 10px rgba(182, 15, 97, 0.9);
    -webkit-transform-origin: bottom center;
  }
  .my-balloons > div:nth-child(1) {
    background: rgba(182, 15, 97, 0.9);
    left: 0;
    top: 7rem;
    box-shadow: inset 10px 10px 10px rgba(135, 11, 72, 0.9);
    -webkit-animation: balloon1 6s ease-in-out infinite;
    -moz-animation: balloon1 6s ease-in-out infinite;
    -o-animation: balloon1 6s ease-in-out infinite;
    animation: balloon1 6s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(1):before {
    color: rgba(182, 15, 97, 0.9);
  }
  .my-balloons > div:nth-child(2) {
    background: rgba(242, 194, 58, 0.9);
    left: 120px;
    top: 4rem;
    box-shadow: inset 10px 10px 10px rgba(192, 154, 43, 0.9);
    -webkit-animation: balloon2 6s ease-in-out infinite;
    -moz-animation: balloon2 6s ease-in-out infinite;
    -o-animation: balloon2 6s ease-in-out infinite;
    animation: balloon2 6s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(2):before {
    color: rgba(242, 194, 58, 0.9);
  }
  .my-balloons > div:nth-child(3) {
    background: rgba(45, 181, 167, 0.9);
    left: 240px;
    box-shadow: inset 10px 10px 10px rgba(35, 140, 129, 0.9);
    -webkit-animation: balloon4 6s ease-in-out infinite;
    -moz-animation: balloon4 6s ease-in-out infinite;
    -o-animation: balloon4 6s ease-in-out infinite;
    animation: balloon4 6s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(3):before {
    color: rgba(45, 181, 167, 0.9);
  }
  .my-balloons > div:nth-child(4) {
    background: rgba(190, 61, 244, 0.9);
    left: 360px;
    top: 4rem;
    box-shadow: inset 10px 10px 10px rgba(160, 31, 204, 0.9);
    -webkit-animation: balloon1 5s ease-in-out infinite;
    -moz-animation: balloon1 5s ease-in-out infinite;
    -o-animation: balloon1 5s ease-in-out infinite;
    animation: balloon1 5s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(4):before {
    color: rgba(190, 61, 244, 0.9);
  }
  .my-balloons > div:nth-child(5) {
    background: rgba(180, 224, 67, 0.9);
    left: 480px;
    top: 7rem;
    box-shadow: inset 10px 10px 10px rgba(140, 179, 47, 0.9);
    -webkit-animation: balloon3 5s ease-in-out infinite;
    -moz-animation: balloon3 5s ease-in-out infinite;
    -o-animation: balloon3 5s ease-in-out infinite;
    animation: balloon3 5s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(5):before {
    color: rgba(180, 224, 67, 0.9);
  }
  .my-balloons > div:nth-child(6) {
    background: rgba(242, 112, 45, 0.9);
    left: 600px;
    box-shadow: inset 10px 10px 10px rgba(202, 92, 35, 0.9);
    -webkit-animation: balloon2 3s ease-in-out infinite;
    -moz-animation: balloon2 3s ease-in-out infinite;
    -o-animation: balloon2 3s ease-in-out infinite;
    animation: balloon2 3s ease-in-out infinite;
  }
  .my-balloons > div:nth-child(6):before {
    color: rgba(242, 112, 45, 0.9);
  }
  .my-balloons > div:before {
    position: absolute;
    bottom: -11px;
    left: 45px;
    content: "â–²";
    font-size: 1em;
  }
  .my-balloons:hover {
    cursor: pointer;
  }
  p {
    font-size: 1em;
    font-weight: bold;
    line-height: 0.5em;
    text-align: center;
    color: white;
    position: relative;
    top: 25%;
  }
  /*BALLOON 1 4*/
  @-webkit-keyframes balloon1 {
    0%,
    100% {
    }
    50% {
      -webkit-transform: translateY(-20px) rotate(8deg);
    }
  }
  @-moz-keyframes balloon1 {
    0%,
    100% {
      -moz-transform: translateY(0) rotate(-6deg);
    }
    50% {
      -moz-transform: translateY(-20px) rotate(8deg);
    }
  }
  @-o-keyframes balloon1 {
    0%,
    100% {
      -o-transform: translateY(0) rotate(-6deg);
    }
    50% {
      -o-transform: translateY(-20px) rotate(8deg);
    }
  }
  @keyframes balloon1 {
    0%,
    100% {
      transform: translateY(0) rotate(-6deg);
    }
    50% {
      transform: translateY(-20px) rotate(8deg);
    }
  }
  /* BAllOON 2 5*/
  @-webkit-keyframes balloon2 {
    0%,
    100% {
      -webkit-transform: translateY(0) rotate(6deg);
    }
    50% {
      -webkit-transform: translateY(-30px) rotate(-8deg);
    }
  }
  @-moz-keyframes balloon2 {
    0%,
    100% {
      -moz-transform: translateY(0) rotate(6deg);
    }
    50% {
      -moz-transform: translateY(-30px) rotate(-8deg);
    }
  }
  @-o-keyframes balloon2 {
    0%,
    100% {
      -o-transform: translateY(0) rotate(6deg);
    }
    50% {
      -o-transform: translateY(-30px) rotate(-8deg);
    }
  }
  @keyframes balloon2 {
    0%,
    100% {
      transform: translateY(0) rotate(6deg);
    }
    50% {
      transform: translateY(-30px) rotate(-8deg);
    }
  }
  /* BAllOON 0*/
  @-webkit-keyframes balloon3 {
    0%,
    100% {
      -webkit-transform: translate(0, -10px) rotate(6deg);
    }
    50% {
      -webkit-transform: translate(-20px, 30px) rotate(-8deg);
    }
  }
  @-moz-keyframes balloon3 {
    0%,
    100% {
      -moz-transform: translate(0, -10px) rotate(6deg);
    }
    50% {
      -moz-transform: translate(-20px, 30px) rotate(-8deg);
    }
  }
  @-o-keyframes balloon3 {
    0%,
    100% {
      -o-transform: translate(0, -10px) rotate(6deg);
    }
    50% {
      -o-transform: translate(-20px, 30px) rotate(-8deg);
    }
  }
  @keyframes balloon3 {
    0%,
    100% {
      transform: translate(0, -10px) rotate(6eg);
    }
    50% {
      transform: translate(-20px, 30px) rotate(-8deg);
    }
  }
  /* BAllOON 3*/
  @-webkit-keyframes balloon4 {
    0%,
    100% {
      -webkit-transform: translate(10px, -10px) rotate(-8deg);
    }
    50% {
      -webkit-transform: translate(-15px, 20px) rotate(10deg);
    }
  }
  @-moz-keyframes balloon4 {
    0%,
    100% {
      -moz-transform: translate(10px, -10px) rotate(-8deg);
    }
    50% {
      -moz-transform: translate(-15px, 10px) rotate(10deg);
    }
  }
  @-o-keyframes balloon4 {
    0%,
    100% {
      -o-transform: translate(10px, -10px) rotate(-8deg);
    }
    50% {
      -o-transform: translate(-15px, 10px) rotate(10deg);
    }
  }
  @keyframes balloon4 {
    0%,
    100% {
      transform: translate(10px, -10px) rotate(-8deg);
    }
    50% {
      transform: translate(-15px, 10px) rotate(10deg);
    }
  }