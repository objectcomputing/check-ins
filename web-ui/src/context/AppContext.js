import React, { useEffect, useReducer, useMemo } from "react";
import { createSelector } from 'reselect';
import { getCurrentUser, updateMember, getAllMembers } from "../api/member";
import { getCheckinByMemberId, createCheckin } from "../api/checkins";
import { BASE_API_URL } from "../api/api";
import axios from "axios";

export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_INDEX = "@@check-ins/update_index";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_CURRENT_CHECKIN = "@@check-ins/update_current_checkin";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const UPDATE_SELECTED_PROFILE = "@@check-ins/update_selected_profile";
const SET_CSRF = "@@check-ins/update_csrf";

const AppContext = React.createContext();

const reducer = (state, action) => {
  switch (action.type) {
    case MY_PROFILE_UPDATE:
      state.userProfile = action.payload;
      break;
    case UPDATE_USER_BIO:
      state.userProfile.memberProfile.bioText = action.payload;
      updateMember(state.userProfile.memberProfile);
      break;
    case UPDATE_CHECKINS:
      state.checkins = action.payload;
      //sort by date
      state.checkins.sort(function (a, b) {
        return new Date(...a.checkInDate) - new Date(...b.checkInDate);
      });
      state.currentCheckin = state.checkins[state.checkins.length - 1];
      break;
    case SET_CSRF:
      state.csrf = action.payload;
      break;
    case UPDATE_TOAST:
      state.toast = action.payload;
      break;
    case UPDATE_TEAMS:
      state.teams = action.payload;
      break;
    case UPDATE_MEMBER_PROFILES:
      state.memberProfiles = action.payload;
      break;
    case UPDATE_TEAM_MEMBERS:
      state.teamMembers = action.payload;
      break;
    case UPDATE_CURRENT_CHECKIN:
      state.currentCheckin = action.payload;
      break;
    case UPDATE_SELECTED_PROFILE:
      const { payload } = action;
      if (state.selectedProfile !== payload) {
        state.selectedProfile = payload;
        state.currentCheckin = payload ? payload.checkIn : {};
      }
      if (payload === undefined) {
        state.checkins = [];
      }
      break;
    default:
  }
  return { ...state };
};

const initialState = {
  checkins: [],
  csrf: undefined,
  currentCheckin: {},
  teams: [],
  memberProfiles: [],
  index: 0,
  selectedProfile: undefined,
  toast: {
    severity: "",
    toast: "",
  },
  userProfile: undefined,
};

const getCheckins = async (id, pdlId, date, dispatch, csrf) => {
  const res = await getCheckinByMemberId(id, csrf);
  let data =
    res.payload && res.payload.data && res.payload.status === 200 && !res.error
      ? res.payload.data
      : null;
  if (data && data.length > 0) {
    const allComplete = data.every((checkin) => checkin.completed === true);
    if (allComplete) {
      const prevCheckinDate = data[data.length - 1].checkInDate;
      if (pdlId) {
        const res = await createCheckin({
          teamMemberId: id,
          pdlId: pdlId,
          checkInDate: date(3, prevCheckinDate),
          completed: false,
        });
        const checkin =
          res.payload && res.payload.data && !res.error
            ? res.payload.data
            : null;
        data.push(checkin);
      }
    }
  } else if (data && data.length === 0) {
    if (pdlId) {
      const res = await createCheckin({
        teamMemberId: id,
        pdlId: pdlId,
        checkInDate: date(1),
        completed: false,
      });
      const checkin =
        res.payload && res.payload.data && !res.error ? res.payload.data : null;
      data = [checkin];
    }
  }
  //without this check you get infinite checkin calls
  if (data.length > 0) {
    dispatch({ type: UPDATE_CHECKINS, payload: data });
  }
};

const AppContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const memberProfile =
    state && state.userProfile && state.userProfile.memberProfile
      ? state.userProfile.memberProfile
      : undefined;
  const id = memberProfile ? memberProfile.id : undefined;
  const selectedProfile = state && state.selectedProfile;
  const selectedId = selectedProfile ? selectedProfile.id : undefined;


  const { csrf } = state;

  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const res = await axios({
          url: `${BASE_API_URL}/csrf/cookie`,
          responseType: "text",
          withCredentials: true,
        });
        if (res && res.data) {
          dispatch({ type: SET_CSRF, payload: res.data });
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf) {
      updateUserProfile();
    }
  }, [csrf]);

  const date = (months, prevCheckinDate) => {
    let currentMonth = new Date().getMonth() + 1;
    let newDate = prevCheckinDate ? new Date(...prevCheckinDate) : new Date();
    if (prevCheckinDate) {
      let prevCheckinMonth = prevCheckinDate.getMonth() + 1;
      newDate.setMonth(
        newDate.getMonth() + currentMonth - prevCheckinMonth >= 3 ? 1 : months
      );
    } else {
      newDate.setMonth(newDate.getMonth() + months);
    }
    const year = newDate.getFullYear();
    const month = newDate.getMonth() + 1;
    const day = newDate.getDate();
    const hours = newDate.getHours();
    const minutes = newDate.getMinutes();
    const dateTimeArray = [year, month, day, hours, minutes, 0];
    return dateTimeArray;
  };

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
      }
    }
    if (csrf) {
      getMemberProfiles();
    }
  }, [csrf]);

  useEffect(() => {
    if (selectedId && csrf) {
      getCheckins(selectedId, id, date, dispatch, csrf);
    }
  }, [csrf, selectedId, id]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export const selectMemberProfiles = state => state.memberProfiles;
export const selectTeamMembers = state => state.teamMembers;

export const selectProfileMap = createSelector(
  selectMemberProfiles,
  (memberProfiles) => {
    if (memberProfiles && memberProfiles.length) {
      memberProfiles = memberProfiles.reduce((mappedById, profile) => {
        mappedById[profile.id] = profile;
        return mappedById;
      }, {});
    }
    return memberProfiles;
  }
);

export const selectMembersByTeamId = id => createSelector(
  selectTeamMembers,
  (teamMembers) => {
    let members = [];
    if (teamMembers && teamMembers.length) {
      members = teamMembers.filter((member) => member.teamid === id);
    }
    return members;
  }
);

export const selectMemberProfilesByTeamId = id => createSelector(
  selectMembersByTeamId(id),
  selectProfileMap,
  (members, profileMap) => members.map((member) => {
      return { ...profileMap[member.memberid], ...member };
    })
);

export { AppContext, AppContextProvider };
