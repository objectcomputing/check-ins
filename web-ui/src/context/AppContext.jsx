import React, { useEffect, useReducer, useMemo } from 'react';
import { reducer, initialState } from './reducer';
import { getCheckins, getAllCheckinsForAdmin } from './thunks';
import {
  MY_PROFILE_UPDATE,
  SET_CSRF,
  SET_ROLES,
  SET_USER_ROLES,
  UPDATE_GUILDS,
  UPDATE_MEMBER_SKILLS,
  UPDATE_MEMBER_PROFILES,
  UPDATE_TERMINATED_MEMBERS,
  UPDATE_SKILLS,
  UPDATE_CERTIFICATIONS,
  UPDATE_TEAMS,
  UPDATE_PEOPLE_LOADING,
  UPDATE_TEAMS_LOADING
} from './actions';
import {
  getCurrentUser,
  getAllMembers,
  getAllTerminatedMembers
} from '../api/member';
import {
  selectCanViewCheckinsPermission,
  selectCanViewTerminatedMembers,
} from './selectors';
import { getAllRoles, getAllUserRoles } from '../api/roles';
import { getMemberSkills } from '../api/memberskill';
import { BASE_API_URL } from '../api/api';
import { getAllGuilds } from '../api/guild';
import { getSkills } from '../api/skill';
import { getAllTeams } from '../api/team';
import {getCertifications} from "../api/certification.js";

const AppContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppContextProvider = props => {
  const [state, dispatch] = useReducer(
    reducer,
    props?.value?.state || initialState
  );
  const userProfile =
    state && state.userProfile ? state.userProfile : undefined;
  const memberProfile =
    userProfile && userProfile.memberProfile
      ? userProfile.memberProfile
      : undefined;

  const id = memberProfile ? memberProfile.id : undefined;
  const pdlId = memberProfile ? memberProfile.pdlId : undefined;
  const {
    csrf,
    guilds,
    teams,
    memberSkills,
    memberProfiles,
    checkins,
    skills,
    certifications,
    roles,
    userRoles
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;
  useEffect(() => {
    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  useEffect(() => {
    async function getGuilds() {
      let res = await getAllGuilds(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_GUILDS, payload: data });
      }
    }
    if (csrf && !guilds) {
      getGuilds();
    }
  }, [csrf, guilds]);

  useEffect(() => {
    async function getTeams() {
      let res = await getAllTeams(csrf);
      let data =
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data) {
        dispatch({ type: UPDATE_TEAMS, payload: data });
        dispatch({ type: UPDATE_TEAMS_LOADING });
      }
    }
    if (csrf && !teams) {
      dispatch({ type: UPDATE_TEAMS_LOADING });
      getTeams();
    }
  }, [csrf, teams, dispatch]);

  useEffect(() => {
    const updateUserProfile = async () => {
      let res = await getCurrentUser(csrf);
      let profile =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profile) {
        dispatch({ type: MY_PROFILE_UPDATE, payload: profile });
      }
    };
    if (csrf && !userProfile) {
      updateUserProfile();
    }
  }, [csrf, userProfile]);

  useEffect(() => {
    const getAllMemberSkills = async () => {
      const res = await getMemberSkills(csrf);
      const memberSkills =
        res && res.payload && res.payload.data ? res.payload.data : null;
      if (memberSkills) {
        dispatch({ type: UPDATE_MEMBER_SKILLS, payload: memberSkills });
      }
    };
    if (csrf && !memberSkills) {
      getAllMemberSkills();
    }
  }, [csrf, memberSkills]);

  useEffect(() => {
    async function getMemberProfiles() {
      let res = await getAllMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_MEMBER_PROFILES, payload: profiles });
        dispatch({ type: UPDATE_PEOPLE_LOADING, payload: false });
      }
    }
    async function getTerminatedMembers() {
      let res = await getAllTerminatedMembers(csrf);
      let profiles =
        res.payload && res.payload.data && !res.error
          ? res.payload.data
          : undefined;

      if (profiles) {
        dispatch({ type: UPDATE_TERMINATED_MEMBERS, payload: profiles });
      }
    }
    if (csrf && userProfile && !memberProfiles) {
      dispatch({ type: UPDATE_PEOPLE_LOADING, payload: true });
      getMemberProfiles();
      if (selectCanViewTerminatedMembers(state)) {
        getTerminatedMembers();
      }
    }
  }, [csrf, userProfile, memberProfiles]);

  useEffect(() => {
    function getAllTheCheckins() {
      if (
        userProfile &&
        userProfile.permissions?.some(p =>
          p?.permission?.includes('CAN_VIEW_CHECKINS_REPORT')
        ) &&
        id &&
        csrf
      ) {
        getAllCheckinsForAdmin(dispatch, csrf);
      } else if (id && csrf && selectCanViewCheckinsPermission(state)) {
        getCheckins(id, pdlId, dispatch, csrf);
      }
    }
    if (csrf && !checkins) {
      getAllTheCheckins();
    }
  }, [csrf, pdlId, id, userProfile, checkins]);

  useEffect(() => {
    const getAllSkills = async () => {
      const res = await getSkills(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_SKILLS, payload: data });
      }
    };
    if (csrf && !skills) {
      getAllSkills();
    }
  }, [csrf, skills]);

  useEffect(() => {
    const getAllCertifications = async () => {
      const res = await getCertifications(csrf);
      const data =
          res &&
          res.payload &&
          res.payload.data &&
          res.payload.status === 200 &&
          !res.error
              ? res.payload.data
              : null;
      if (data && data.length > 0) {
        dispatch({ type: UPDATE_CERTIFICATIONS, payload: data });
      }
    };
    if (csrf && !certifications) {
      getAllCertifications();
    }
  }, [csrf, certifications]);

  useEffect(() => {
    const getRoles = async () => {
      const res = await getAllRoles(csrf);
      const data =
        res &&
        res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
          ? res.payload.data
          : null;
      if (data && Array.isArray(data) && data.length > 0) {
        dispatch({ type: SET_ROLES, payload: data });
      }
    };
    if (csrf && !roles) {
      getRoles();
    }
  }, [csrf, roles]);

  useEffect(() => {
    const getUserRoles = async () => {
      // make call to the API
      let res = await getAllUserRoles(csrf);
      return res.payload &&
        res.payload.data &&
        res.payload.status === 200 &&
        !res.error
        ? res.payload.data
        : null;
    };

    if (csrf && !userRoles) {
      getUserRoles().then(userRoles => {
        dispatch({ type: SET_USER_ROLES, payload: userRoles });
      });
    }
  }, [csrf, userRoles]);

  const value = useMemo(() => {
    return { state, dispatch };
  }, [state]);
  return (
    <AppContext.Provider value={props.value || value}>
      {props.children}
    </AppContext.Provider>
  );
};

export { AppContext, AppContextProvider };
