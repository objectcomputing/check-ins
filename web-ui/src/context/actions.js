export const MY_PROFILE_UPDATE = "@@check-ins/update_profile";
export const UPDATE_USER_BIO = "@@check-ins/update_bio";
export const UPDATE_CHECKINS = "@@check-ins/update_checkins";
export const UPDATE_CHECKIN = "@@check-ins/update_checkin";
export const ADD_CHECKIN = "@@check-ins/add_checkin";
export const UPDATE_TOAST = "@@check-ins/update_toast";
export const UPDATE_MEMBER_SKILLS = "@@check-ins/update_member_skills";
export const DELETE_MEMBER_SKILL = "@@check-ins/delete_member_skill";
export const ADD_MEMBER_SKILL = "@@check-ins/add_member_skill";
export const UPDATE_TEAMS = "@@check-ins/update_teams";
export const UPDATE_MEMBER_PROFILES = "@@check-ins/update_member_profiles";
export const UPDATE_TEAM_MEMBERS = "@@check-ins/update_team_members";
export const ADD_SKILL = "@@check-ins/add-skill";
export const DELETE_SKILL = "@@check-ins/delete-skill";
export const UPDATE_SKILL = "@@check-ins/update_skill";
export const UPDATE_SKILLS = "@@check-ins/update_skills";
export const ADD_TEAM = "@@check-ins/add_team";
export const SET_CSRF = "@@check-ins/update_csrf";
export const SET_ROLES = "@@check-ins/set_roles";
