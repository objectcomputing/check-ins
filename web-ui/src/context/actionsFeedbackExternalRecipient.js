export const SET_CSRF = '@@check-ins/update_csrf';
export const UPDATE_TOAST = '@@check-ins/update_toast';

