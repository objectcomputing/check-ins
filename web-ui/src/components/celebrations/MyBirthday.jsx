import React, { useContext } from "react";
import { AppContext } from "../../context/AppContext";

import { selectCurrentUser } from "../../context/selectors";

import "./MyBirthday.css";
import "./Birthdays.css";

const MyBirthday = () => {
  const { state } = useContext(AppContext);

  let me = selectCurrentUser(state);

  return (
    <div class="my-birthday-container">
      <div class="my-balloons">
        <div>
          <p>Happy</p>
          <p>Birthday!</p>
        </div>
        <div>
          <p>Happy</p>
          <p>Birthday!</p>
        </div>
        <div>
          <p>Happy</p>
          <p>Birthday!</p>
        </div>
        <div>
          <p>Happy</p>
          <p>Birthday!</p>
        </div>
        <div>
          <p>Happy</p>
          <p>Birthday!</p>
        </div>
      </div>
      <div className="my-birthday">
        <h1>Happy Birthday {me.firstName}!!!</h1>
      </div>
    </div>
  );
};

export default MyBirthday;
