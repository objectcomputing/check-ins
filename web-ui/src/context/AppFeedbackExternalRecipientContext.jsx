import React, { useEffect, useReducer, useMemo } from 'react';
import {reducerFeedbackExternalRecipient, initialState} from "./reducerFeedbackExternalRecilpients.js";
import {BASE_API_URL} from "../api/apiFeedbackExternalRecipient.js";
import {SET_CSRF} from "./actionsFeedbackExternalRecipient.js";

const AppFeedbackExternalRecipientContext = React.createContext();

function getSessionCookieValue(name) {
  const cookies = document?.cookie?.split(';');
  for (let i = 0; i < cookies.length; i++) {
    const cookie = cookies[i].trim();
    if (cookie.startsWith(name + '=')) {
      return decodeURIComponent(cookie.substring(name.length + 1));
    }
  }
  return null;
}

const AppFeedbackExternalRecipientContextProvider = props => {
  console.log("AppFeedbackExternalRecipientContextProvider props: ", props);
  const [state, dispatch] = useReducer(
      reducerFeedbackExternalRecipient,
    props?.value?.state || initialState
  );

  const {
    csrf,
  } = state;
  const url = `${BASE_API_URL}/csrf/cookie`;

  useEffect(() => {

    console.log("AppFeedbackExternalRecipientContextProvider useEffect, csrf");

    const getCsrf = async () => {
      if (!csrf) {
        const payload = getSessionCookieValue('_csrf');
        if (payload) {
          dispatch({ type: SET_CSRF, payload });
        } else {
          const res = await fetch(url, {
            responseType: 'text',
            credentials: 'include'
          });
          if (res && res.ok) {
            dispatch({ type: SET_CSRF, payload: await res.text() });
          }
        }
      }
    };
    getCsrf();
  }, [csrf]);

  const value = useMemo(() => {
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, state: ", state);
    console.log("AppFeedbackExternalRecipientContextProvider useMemo, dispatch: ", dispatch);
    return { state, dispatch };
  }, [state]);

  return (
    <AppFeedbackExternalRecipientContext.Provider value={props.value || value}>
      {props.children}
    </AppFeedbackExternalRecipientContext.Provider>
  );
};

export { AppFeedbackExternalRecipientContext, AppFeedbackExternalRecipientContextProvider };
