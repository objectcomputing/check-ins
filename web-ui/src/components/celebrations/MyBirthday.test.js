import React from "react";
import MyBirthday from "./MyBirthday";
import { AppContextProvider } from "../../context/AppContext";
import { BrowserRouter } from "react-router-dom";

const me = {
  "name": "SumanMaroju",
  "birthDay": "12/27",
  "userId": "1b4f99da-ef70-4a76-9b37-8bb783b749ad"
}

it("renders correctly", () => {
  snapshot(
    <AppContextProvider>
      <BrowserRouter>
        <MyBirthday
          me={me}
        />
      </BrowserRouter>
    </AppContextProvider>
  );
});
