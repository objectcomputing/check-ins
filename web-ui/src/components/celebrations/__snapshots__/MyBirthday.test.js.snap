// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="my-birthday-container"
>
  <div
    className="my-balloons"
    onClick={[Function]}
  >
    <div>
      <p>
        Happy
      </p>
      <p>
        Birthday!
      </p>
    </div>
    <div>
      <p>
        Happy
      </p>
      <p>
        Birthday!
      </p>
    </div>
    <div>
      <p>
        Happy
      </p>
      <p>
        Birthday!
      </p>
    </div>
    <div>
      <p>
        Happy
      </p>
      <p>
        Birthday!
      </p>
    </div>
    <div>
      <p>
        Happy
      </p>
      <p>
        Birthday!
      </p>
    </div>
  </div>
  <div
    className="my-birthday"
  >
    <h1>
      Happy Birthday 
      !!!
    </h1>
  </div>
</div>
`;
